pasture-algorithms/src/normal_estimation.rs

```diff
@@ -0,0 +1,551 @@
+        fn at(&self, k: usize) -> f64 {
+            return self.position[k];
+        }
+    }
+
+    #[test]
+    fn test_compute_normal_sub() {
+        let points: Vec<SimplePoint> = vec![
+            SimplePoint {
+                position: Vector3::new(1.0, 0.0, 0.0),
+                intensity: 42,
+            },
+            SimplePoint {
+                position: Vector3::new(0.0, 1.0, 0.0),
+                intensity: 84,
+            },
+            SimplePoint {
+                position: Vector3::new(1.0, 1.0, 0.0),
+                intensity: 84,
+            },
+            SimplePoint {
+                position: Vector3::new(-1.0, 0.0, 0.0),
+                intensity: 84,
+            },
+        ];
+
+        let mut interleaved = InterleavedVecPointStorage::new(SimplePoint::layout());
+
+        interleaved.push_points(points.as_slice());
+
+        let mut centroid = Vector4::<f64>::new_random();
+        compute_centroid(&interleaved, &mut centroid);
+        let result_centroid = Vector4::new(0.25, 0.5, 0.0, 1.0);
+        assert_eq!(centroid, result_centroid);
+
+        let mut covariance_matrix = DMatrix::<f64>::zeros(3, 3);
+        compute_covarianz_matrix(&interleaved, &mut covariance_matrix);
+        let result = Matrix3::new(
+            0.6875 * 4.0,
+            0.125 * 4.0,
+            0.0,
+            0.125 * 4.0,
+            0.25 * 4.0,
+            0.0,
+            0.0,
+            0.0,
+            0.0,
+        );
+        assert_eq!(covariance_matrix, result);
+
+        let mut normal_vec = vec![0.0; 3];
+        let mut curvature: f64 = 0.0;
+
+        solve_plane_parameter(&covariance_matrix, &mut normal_vec, &mut curvature);
+
+        assert_eq!(normal_vec[0], 0.0);
+        assert_eq!(normal_vec[1], 0.0);
+        assert_ne!(normal_vec[2], 0.0);
+        assert_eq!(curvature, 0.0);
+    }
+
+    #[test]
+    fn test_compute_normal() {
+        let points: Vec<SimplePoint> = vec![
+            SimplePoint {
+                position: Vector3::new(1.0, 0.0, 0.0),
+                intensity: 42,
+            },
+            SimplePoint {
+                position: Vector3::new(0.0, 1.0, 0.0),
+                intensity: 84,
+            },
+            SimplePoint {
+                position: Vector3::new(1.0, 1.0, 0.0),
+                intensity: 84,
+            },
+            SimplePoint {
+                position: Vector3::new(-1.0, 0.0, 0.0),
+                intensity: 84,
+            },
+        ];
+
+        let mut interleaved = InterleavedVecPointStorage::new(SimplePoint::layout());
+
+        interleaved.push_points(points.as_slice());
+
+        let solution_vec =
+            compute_normals::<InterleavedVecPointStorage, SimplePoint>(&interleaved, 3);
+        for solution in solution_vec {
+            assert_eq!(solution.1, 0.0);
+            assert_eq!(solution.2, 0.0);
+            assert_ne!(solution.3, 0.0);
+            assert_eq!(solution.4, 0.0);
+        }
+    }
+}
```

Comment:
You should also have some tests for the error cases, e.g. passed in too few points or some points that don't have a normal.